File: src/investments/registered-education-savings-plan/income-level.ts
/*
 Sources
 https://www.canada.ca/en/services/benefits/education/education-savings/estimating-amounts.html#_clb_income

 Last updated
 2025-12-21
 */

const LOW_INCOME_THRESHOLD = 57375;
const MEDIUM_INCOME_THRESHOLD = 114750;

export enum IncomeLevelType {
    LOW = 'LOW',
    MEDIUM = 'MEDIUM',
    HIGH = 'HIGH'
}

export interface IncomeLevel {
    LOW_INCOME_THRESHOLD: number;
    MEDIUM_INCOME_THRESHOLD: number;
    getIncomeLevel(income: number): IncomeLevelType;
}

const getIncomeLevel = (income: number): IncomeLevelType => {
    if (income <= LOW_INCOME_THRESHOLD) {
        return IncomeLevelType.LOW;
    }

    if (income <= MEDIUM_INCOME_THRESHOLD) {
        return IncomeLevelType.MEDIUM;
    }

    return IncomeLevelType.HIGH;
};

export const IncomeLevel: IncomeLevel = {
    LOW_INCOME_THRESHOLD,
    MEDIUM_INCOME_THRESHOLD,
    getIncomeLevel,
};

File: src/commonMain/kotlin/com/equisoft/taxca/investments/resp/IncomeLevel.kt
/*
 Sources
 https://www.canada.ca/en/services/benefits/education/education-savings/estimating-amounts.html#_clb_income

 Last updated
 2025-12-21
 */

package com.equisoft.taxca.investments.resp

enum class IncomeLevelType {
    LOW,
    MEDIUM,
    HIGH,
}

// The TS source merges an interface and a const named `IncomeLevel`; Kotlin forbids a
// top-level class and val sharing a name, so the class carries a `Resp` prefix.
data class RespIncomeLevel(
    val lowIncomeThreshold: Double,
    val mediumIncomeThreshold: Double,
) {
    fun getIncomeLevel(income: Double): IncomeLevelType {
        if (income <= lowIncomeThreshold) {
            return IncomeLevelType.LOW
        }

        if (income <= mediumIncomeThreshold) {
            return IncomeLevelType.MEDIUM
        }

        return IncomeLevelType.HIGH
    }
}

val IncomeLevel: RespIncomeLevel = RespIncomeLevel(
    lowIncomeThreshold = 57375.0,
    mediumIncomeThreshold = 114750.0,
)
